Core/src/IfcGeometryConverter/CurveConverter.h

```diff
@@ -279,51 +264,24 @@ namespace OpenInfraPlatform {
 					} // end if IfcBSplineCurve
 
 					// (3/6) IfcCompositeCurve SUBTYPE OF IfcBoundedCurve
-					std::shared_ptr<typename IfcEntityTypesT::IfcCompositeCurve> composite_curve =
-						std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcCompositeCurve>(bounded_curve.lock());
-					if (composite_curve) {
-
-						std::vector<std::shared_ptr<typename IfcEntityTypesT::IfcCompositeCurveSegment> > segments;
-						segments.resize(composite_curve->Segments.size());
-
-						std::transform(
-							composite_curve->Segments.begin(),
-							composite_curve->Segments.end(),
-							segments.begin(),
-							[](auto &it) { return it.lock(); });
-
-						std::vector<std::shared_ptr<typename IfcEntityTypesT::IfcCompositeCurveSegment> >::iterator it_segments =
-							segments.begin();
-
-						for (; it_segments != segments.end(); ++it_segments) {
-							std::shared_ptr<typename IfcEntityTypesT::IfcCompositeCurveSegment> segment = (*it_segments);
-							std::shared_ptr<typename IfcEntityTypesT::IfcCurve> segment_curve = segment->ParentCurve.lock();
-
-							std::vector<carve::geom::vector<3>> segment_vec;
-							convertIfcCurve(segment_curve, segment_vec, segmentStartPoints);
-							if (!segment_vec.empty()) {
-								GeomUtils::appendPointsToCurve(segment_vec, targetVec);
-							}
-						}
-						return;
+					else if (bounded_curve.isOfType<typename IfcEntityTypesT::IfcCompositeCurve>()) {
+						return convertIfcCompositeCurve(bounded_curve.as<typename IfcEntityTypesT::IfcCompositeCurve>(), targetVec, segmentStartPoints);
 					} // end if IfcCompositeCurve
 
 					// (4/6) IfcIndexedPolyCurve SUBTYPE OF IfcBoundedCurve
-					if (bounded_curve.isOfType<typename IfcEntityTypesT::IfcIndexedPolyCurve>())
+					else if (bounded_curve.isOfType<typename IfcEntityTypesT::IfcIndexedPolyCurve>())
 					{
 						return convertIfcIndexedPolyCurve(
 							bounded_curve.as<typename IfcEntityTypesT::IfcIndexedPolyCurve>(),
 							targetVec, segmentStartPoints,
 							trim1Vec, trim2Vec, senseAgreement
 						);
-					}
+					} // end if IfcIndexedPolyCurve
 
 					// (5/6) IfcPolyline SUBTYPE OF IfcBoundedCurve
-					std::shared_ptr<typename IfcEntityTypesT::IfcPolyline> poly_line =
-						std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcPolyline>(bounded_curve.lock());
-					if (poly_line) {
-						if (!poly_line->Points.empty()) {
-							std::vector<carve::geom::vector<3>> loop = convertIfcPolyline(EXPRESSReference<typename IfcEntityTypesT::IfcPolyline>(poly_line));
+					else if (bounded_curve.isOfType<typename IfcEntityTypesT::IfcPolyline>()) {
+						if (!bounded_curve.as<typename IfcEntityTypesT::IfcPolyline>()->Points.empty()) {
+							std::vector<carve::geom::vector<3>> loop = convertIfcPolyline(EXPRESSReference<typename IfcEntityTypesT::IfcPolyline>(bounded_curve.as<typename IfcEntityTypesT::IfcPolyline>()));

```

Comment:
```suggestion
							std::vector<carve::geom::vector<3>> loop = convertIfcPolyline(bounded_curve.as<typename IfcEntityTypesT::IfcPolyline>());
```

`.as<...>` already returns an EXPRESSReference, no need to produce a copy (at least I hope so).

Reply:
Resolved:  4878063